serial/aio.py

```diff
@@ -1,121 +1,127 @@
-        self._loop.call_soon(self._protocol.connection_lost, exc)
-
-    def _read_ready(self):
-        try:
-            data = self.serial.read(1024)
-        except serial.SerialException as e:
-            self.close(exc=e)
-        else:
-            if data:
-                self._protocol.data_received(data)
-
-    def write(self, data):
-        try:
-            self.serial.write(data)
-        except serial.SerialException as e:
-            self.close(exc=e)
-
-    def can_write_eof(self):
-        return False
-
-    def pause_reading(self):
-        if self._closing:
-            raise RuntimeError('Cannot pause_reading() when closing')
-        if self._paused:
-            raise RuntimeError('Already paused')
-        self._paused = True
-        self._loop.remove_reader(self._sock_fd)
-        if self._loop.get_debug():
-            logging.debug("%r pauses reading", self)
-
-    def resume_reading(self):
-        if not self._paused:
-            raise RuntimeError('Not paused')
-        self._paused = False
-        if self._closing:
-            return
-        self._loop.add_reader(self._sock_fd, self._read_ready)
-        if self._loop.get_debug():
-            logging.debug("%r resumes reading", self)
-
-    #~ def set_write_buffer_limits(self, high=None, low=None):
-    #~ def get_write_buffer_size(self):
-    #~ def writelines(self, list_of_data):
-    #~ def write_eof(self):
-    #~ def abort(self):
-
-
-@asyncio.coroutine
-def create_serial_connection(loop, protocol_factory, *args, **kwargs):
-    ser = serial.Serial(*args, **kwargs)
-    protocol = protocol_factory()
-    transport = SerialTransport(loop, protocol, ser)
-    return (transport, protocol)
-
-# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
-# test
-if __name__ == '__main__':
-    class Output(asyncio.Protocol):
-        def connection_made(self, transport):
-            self.transport = transport
-            print('port opened', transport)
-            transport.serial.rts = False
-            transport.write(b'hello world\n')
-
-        def data_received(self, data):
-            print('data received', repr(data))
-            self.transport.close()
-
-        def connection_lost(self, exc):
-            print('port closed')
-            asyncio.get_event_loop().stop()
-
-    loop = asyncio.get_event_loop()
-    coro = create_serial_connection(loop, Output, '/dev/ttyUSB0', baudrate=115200)
-    loop.run_until_complete(coro)
-    loop.run_forever()
-    loop.close()
+
+        self.read_future = None
+        self.write_future = None
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    async def read_exactly(self, n):
+        data = bytesarray()
+        while len(data) < n:
+            remaining = n - len(data)
+            data += await self.read(remaining)
+        return data
+
+    async def write_exactly(self, data):
```

Comment:
isn't `async def` python 3.5 only?  It's still possible to do with the decorators right?